File: WCPhotoManipulator/FileUtils.m
//
//  FileUtils.m
//  WCPhotoManipulator
//
//  Created by Woraphot Chokratanasombat on 24/3/19.
//  Copyright © 2019 Woraphot Chokratanasombat. All rights reserved.
//

#import "FileUtils.h"
#import "MimeUtils.h"

@implementation FileUtils

+ (NSString *)createTempFile:(NSString *)prefix mimeType:(NSString *)mimeType {
    NSString *extension = [MimeUtils toExtension:mimeType];
    NSString *fileName = [NSString stringWithFormat:@"%@%@%@", prefix, [[NSUUID UUID] UUIDString], extension];
    return [FileUtils.cachePath stringByAppendingPathComponent:fileName];
}

+ (NSString *)cachePath {
    NSArray *array = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES);
    return [array objectAtIndex:0];
}

+ (void)cleanDirectory:(NSString *)path prefix:(NSString *)prefix {
    NSArray *files = [FileUtils filesIn:path withPrefix:prefix];
    NSFileManager *fileManager = [NSFileManager defaultManager];
    
    [files enumerateObjectsUsingBlock:^(NSString *file, __unused NSUInteger idx, __unused BOOL *stop) {
        [fileManager removeItemAtPath:file error:nil];
    }];
}
+ (NSArray *)filesIn:(NSString *)path withPrefix:(NSString *)prefix {
    NSArray *files = [[NSFileManager defaultManager] contentsOfDirectoryAtPath:path error:nil];
    files = [files filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"self BEGINSWITH[cd] %@", prefix]];
    
    NSMutableArray *result = [NSMutableArray array];
    [files enumerateObjectsUsingBlock:^(NSString *name, __unused NSUInteger idx, __unused BOOL *stop) {
        [result addObject:[path stringByAppendingPathComponent:name]];
    }];
    return result;
}

+ (void)saveImageFile:(UIImage *)image mimeType:(NSString *)mimeType quality:(CGFloat)quality file:(NSString *)file {
    NSData *data = [FileUtils imageToData:image mimeType:mimeType quality:quality];
    
    [data writeToFile:file atomically:YES];
}

+ (NSData *)imageToData:(UIImage *)image mimeType:(NSString *)mimeType quality:(CGFloat)quality {
    if ([MimeUtils.PNG isEqual:mimeType]) return UIImagePNGRepresentation(image);
    else return UIImageJPEGRepresentation(image, quality / 100);
}

+ (UIImage *)imageFromUrl:(NSURL *)url {
    if ([url.scheme isEqual:@"file"]) return [UIImage imageWithContentsOfFile:url.path];
    
    NSData *data = [[NSData alloc] initWithContentsOfURL:url];
    return [UIImage imageWithData:data];
}

@end

File: WCPhotoManipulator/FileUtils.swift
//
//  FileUtils.swift
//  WCPhotoManipulator
//
//  Created by Woraphot Chokratanasombat on 28/7/20.
//  Copyright © 2020 Woraphot Chokratanasombat. All rights reserved.
//

import Foundation
import UIKit

@objc public class FileUtils: NSObject {
    @objc public class func createTempFile(_ prefix: String, mimeType: String) -> String {
        let ext = MimeUtils.toExtension(mimeType)
        let filename = "\(prefix)\(UUID().uuidString)\(ext)"
        return (cachePath() as NSString).appendingPathComponent(filename)
    }
    
    @objc public class func cachePath() -> String {
        return NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true)[0]
    }
    
    @objc public class func cleanDirectory(_ path: String, prefix: String) {
        let fileManager = FileManager.default
        
        filesIn(path, withPrefix: prefix).forEach { it in
            try? fileManager.removeItem(atPath: it)
        }
    }
    
    @objc internal class func filesIn(_ path: String, withPrefix prefix: String) -> [String] {
        return (try? FileManager.default
            .contentsOfDirectory(atPath: path)
            .filter({it in it.starts(with: prefix)})
            .map({it in (path as NSString).appendingPathComponent(it)}))
        ?? []
    }
    
    @objc public class func saveImageFile(_ image: UIImage, mimeType: String, quality: CGFloat, file: String) {
        let data = imageToData(image, mimeType:mimeType, quality:quality)
        
        try? data?.write(to: URL(fileURLWithPath: file), options: .atomic)
    }

    @objc public class func imageFromUrl(_ url: URL) -> UIImage? {
        if url.scheme == "file" {
            return UIImage(contentsOfFile: url.path)
        }
        return UIImage(data: try! Data(contentsOf: url))
    }

    @objc internal class func imageToData(_ image: UIImage, mimeType: String, quality: CGFloat) -> Data? {
        if (MimeUtils.PNG == mimeType) {
            return image.pngData()
        }
        return image.jpegData(compressionQuality: quality / 100)
    }
}
